### PR TITLE
valdiate: ensure getfields works with interface{}

### PR DESCRIPTION
**Pull request by ajeddeloh**
Ensure getFields() will correctly unbox an embedded interface{} that
actually contains a struct type. This is used in the config transpiler.

### CODE REVIEW

**Comment by ajeddeloh** [comment]
I'll add some tests.

This is for this PR, specifically the Etcd and Flannel structs.
https://github.com/coreos/container-linux-config-transpiler/pull/18

Etcd//Flannel needs an embedded empty interface. It needs to be embedded to ensure the yaml tree structure and the parsed tree structure are the same. It needs to be an interface{} to ensure the options can be swapped out for either Etcd3 or Etcd2 structs. Go disallows embedded a interface{} directly, but you can embed an alias to one.

The problem is Go's reflection library isn't very consistent.

Normally `reflect.TypeOf(x)` or `refelct.ValueOf(x).Type()` will tell you the underlying type of an interface{} or alias to interface{}. It appears that it does not when you use a `reflect.Value` to get another `reflect.Value` via something like the `Field()` method.

See here: https://play.golang.org/p/jEUYj_9a93

In short, the `reflect.ValueOf(v.Interface())` gets rid of the type alias and gets the actual type the interface{} is holding, regardless of how the `reflect.Value` was obtained. It's... not intuitive.

**Comment by crawford** [comment]
Can you throw that explanation into the commit message as well as the internal golang-bugs document?
